ovos_dinkum_listener/_util.py

```diff
@@ -0,0 +1,123 @@
+
+    Each instance of this builder can be customized to support different keys,
+    but some common ones are builtin like "uuid4", "now", and "utcnow"
+
+    Example:
+        >>> # Simple now and uuid4 keys are available by default.
+        >>> template = 'my_filename_{now}_{uuid4}'
+        >>> self = _TemplateFilenameFormatter()
+        >>> name = self.format(template)
+        >>> # xdoctest: +IGNORE_WANT
+        >>> print(f'name={name}')
+        name=my_filename_2024-09-14 18:53:22.619838-05:00_7fe91270-3266-42c1-89d9-0809b9facb9e
+
+    Example:
+        >>> # The now can use standard python format-string semantics
+        >>> template = 'my_filename_{now:%Y-%m-%dT%H%M%S%z}_{uuid4}'
+        >>> self = _TemplateFilenameFormatter()
+        >>> name = self.format(template)
+        >>> # xdoctest: +IGNORE_WANT
+        >>> print(f'name={name}')
+        name=my_filename_2024-09-14T185354-0500_6f0f6daf-cd81-4c5b-bf38-76a4466161c6
+
+    Example:
+        >>> # You can define how to handle custom keys
+        >>> template = '{mykey}.bar.{now:%Y-%z}-{uuid4}'
+        >>> self = _TemplateFilenameFormatter()
+        >>> @self.register('mykey')
+        >>> def custom_func():
+        ...     return 'myval'
+        >>> name = self.format(template)
+        >>> # xdoctest: +IGNORE_WANT
+        >>> print(f'name={name}')
+        name=myval.bar.2024--765176fa-7c80-431c-b43d-2ad14a58a249
+
+    Example:
+        >>> # should raise an error if template contains an unknown field
+        >>> template = '{doesnotexist}.bar.{now:%Y-%z}-{uuid4}'
+        >>> self = _TemplateFilenameFormatter()
+        >>> import pytest
+        >>> with pytest.raises(KeyError) as ex:
+        ...     name = self.format(template)
+        >>> # xdoctest: +IGNORE_WANT
+        >>> print(str(ex.value))
+        "Template string contained unsupported keys ['doesnotexist']. Supported keys are: ['uuid4', 'now', 'utcnow']"
+
+    """
+    def __init__(self):
+        import uuid
+        # import datetime as datetime_mod
+        # mapping of key to functions that build content for those keys
+        self.builders = {
+            'uuid4': uuid.uuid4,
+            'now': self._now_tz,
+            'utcnow': self._utcnow_tz,
+        }
+
+    def register(self, key):
+        """
+        Decorator which will register a function called when the template
+        string contains ``key``.
+        """
+        def _decor(func):
+            self.builders[key] = func
+            return func
+        return _decor
+
+    def _now_tz(self):
+        # Helper to provide a timzone aware datetime.now variant.
+        # (note: the timezone is not always right, e.g. EDT-vs-EST)
+        import datetime as datetime_mod
+        import time
+        _delta = datetime_mod.timedelta(seconds=-time.timezone)
+        tzinfo = datetime_mod.timezone(_delta)
+        datetime_obj = datetime_mod.datetime.now(tzinfo)
+        return datetime_obj
```

Comment:
should use `now_local()` instead https://github.com/OpenVoiceOS/ovos-utils/blob/dev/ovos_utils/time.py#L33